src/wallet/wallet.cpp

```diff
@@ -7065,36 +7071,94 @@ bool SpendableInputs::LimitToAmount(const CAmount amountRequired, const CAmount
     }
     sproutNoteEntries.erase(sproutIt, sproutNoteEntries.end());
 
-    // Next select transparent utxos. We preferentially spend transparent funds,
-    // with the intent that we'd like to opportunistically shield whatever is
-    // possible, and we will always shield change after the introduction of
-    // unified addresses.
-    std::sort(utxos.begin(), utxos.end(),
-        [](COutput i, COutput j) -> bool {
-            return i.Value() > j.Value();
-        });
-    auto utxoIt = utxos.begin();
-    while (utxoIt != utxos.end() && !haveSufficientFunds()) {
-        totalSelected += utxoIt->Value();
-        ++utxoIt;
-    }
-    utxos.erase(utxoIt, utxos.end());
-
-    // Finally select Sapling outputs. After the introduction of Orchard to the
-    // wallet, the selection of Sapling and Orchard notes, and the
-    // determination of change amounts, should be done in a fashion that
-    // minimizes information leakage whenever possible.
-    std::sort(saplingNoteEntries.begin(), saplingNoteEntries.end(),
-        [](SaplingNoteEntry i, SaplingNoteEntry j) -> bool {
-            return i.note.value() > j.note.value();
-        });
-    auto saplingIt = saplingNoteEntries.begin();
-    while (saplingIt != saplingNoteEntries.end() && !haveSufficientFunds()) {
-        totalSelected += saplingIt->note.value();
-        ++saplingIt;
+    // Check what input pools we have available.
+    bool haveTransparent = !utxos.empty();
+    bool haveSapling = !saplingNoteEntries.empty();
+    std::set<OutputPool> available;
+    if (haveTransparent) {
+        available.insert(OutputPool::Transparent);
+    }
+    if (haveSapling) {
+        available.insert(OutputPool::Sapling);
+    }
+
+    // Now determine the order in which to select the remaining notes and coins.
+    // We do this in a way that minimizes information leakage while moving funds
+    // into the shielded pool where possible.
+    //
+    // In the following table:
+    // - T: transparent pool
+    // - S: Sapling pool
+    //
+    // Available | Recipients | Order | Rationale
+    // ----------|------------|-------|----------
+    //    T      |    T       |  T    | N/A
+    //    T      |     S      |  T    | N/A
+    //    T      |    TS      |  T    | N/A
+    //     S     |    T       |  S    | N/A
+    //     S     |     S      |  S    | N/A
+    //     S     |    TS      |  S    | N/A
+    //    TS     |    T       |  ST   | Hide sender if possible.
+    //    TS     |     S      |  TS   | Opportunistic shielding.
+    //    TS     |    TS      |  TS   | Opportunistic shielding.
+    std::vector<OutputPool> selectionOrder;
+    if (available.size() == 1) {
+        // We only have one input pool, and thus only one pool change can go to.
+        selectionOrder.assign(available.begin(), available.end());
+    } else if (recipientPools.size() == 1 && recipientPools.count(OutputPool::Transparent)) {
+        // Hide sender if possible.
```

Comment:
This comment should be expanded a bit to explain the scenario, because it's a little subtle.